### DISCUSSION TITLE
Wildcard subdomains in preview URLs?

### DISCUSSION THREAD

**Discussion by lardissone** [vercel/community#475] [Help] [answered]
**Is there a way to use wildcard subdomains on preview URLs?**

We need our dev team to add comments to our PRs before we merge to staging branch.
So that they should be able to see the site working somewhere to do the corresponding QA.
The problem is that our current multi-tenant app uses subdomain to know which tenant is working on, but preview URLs generated on PRs seems to not allow subdomains.

I know that probably a solution would be use a new branch for testing, but since Vercel offers this awesome feature it would be easier and our development would be faster too because we can test several things in parallel.

**Comment by paulogdm** [comment, marked as answer]
Perhaps you are looking for the "preview deployment suffix" add-on, which replaces "vercel.app" with an arbitrary domain for development purposes? If so, you can enable it on your billing page.

**Reply by lardissone** [reply]
But you can use subdomains with it the same way you do with custom domains?

Anyway no, I need subdomains to be allowed to be used without pre-definition. So that if there's a new tenant in the database with subdomain `tenant1` I could access something like: `tenant1.git-branch-something-124.org.vercel.app` or similar.

**Reply by paulogdm** [reply]
> But you can use subdomains with it the same way you do with custom domains?

Yes, you can.

> Anyway no, I need subdomains to be allowed to be used without pre-definition.

Can you elaborate?

**Reply by lardissone** [reply]
The thing is that I cannot predefine the subdomains since they're created on the fly based on user registrations, I need wildcard subdomains, is that possible with the Preview Deployment Suffix?

**Reply by paulogdm** [reply]
Perhaps one of the following approaches can assist you:

1. `vercel alias --help`
2. https://vercel.com/docs/api#endpoints/aliases/assign-an-alias-to-a-deployment

**Reply by isaachinman** [reply]
@lardissone Did you arrive at a solution for this? I am facing the same set of requirements, and it seems the `Preview Deployment Suffix` feature makes no mention of adding custom DNS records.

@paulogdm Was your suggestion in this case to manually perform a call to `POST /v2/now/deployments/:id/aliases` after each build within the Vercel platform, to be able to use subdomains on preview deployment URLs?

**Reply by lardissone** [reply]
@isaachinman no luck on this yet. A solution was to create a github action to call this `vercel alias` command but it only works for known subdomains.

**Reply by rskvazh** [reply]
Same problem here, we want to test SaaS multiple tenants on preview URLs. We only can do it with production url like `*.mysaas.com`, but not for preview deployments

**Reply by paulogdm** [reply]
You can assign a wildcard domain to a git branch and test it there. Eg:


| Branch        | Domains
| ------------- |:-------------:|
| main     | `*.mysaas.com`
| staging | `*.staging.mysaas.com`

Another idea can be to accept a querystring as an argument for the domain if you are not in production. There are several options atm.

**Reply by april-duong** [reply]
@paulogdm In the example above, assigning `*.staging.mysaas.com` would require us to change the nameservers of the apex domain `mysass.com` according to the docs (https://vercel.com/docs/custom-domains#wildcard-domains), but if we are utilizing the existing nameservers that would cause problems. Is there a different solution to this?

**Reply by paulogdm** [reply]
You can use your own Nameservers, but then you need to upload a custom certificate, which is only available in the Enterprise offering at the moment.

**Reply by gregonarash** [reply]
I think this issue is still no resolved 

docs for https://github.com/vercel/platforms  suggest that:

> you can use wildcard subdomains on .vercel.app links that are associated with your Vercel team slug

Testing however it does not seem that you can deploy Platforms on platforms-myfork.vercel.app  and expect the *.platforms-myfork.vercel.app to work.  Running into issues with SSL.

**Reply by claudiowilson** [reply]
@gregonarash I can confirm that I am also seeing this issue

**Reply by amyegan** [reply]
It should work if you add the `*.<user|team>.vercel.app` wildcard domain to your project settings. Note that this only works with your user name for hobby accounts or team name for teams ([middleware.ts#L33](https://github.com/vercel/platforms/blob/3246e5a2dae8c7686254d1f6115826166aba2806/middleware.ts#L33)). It won't work with the [generated URL](https://vercel.com/docs/concepts/deployments/generated-urls#) because of SSL cert issues, as you saw.

**Reply by StanBoyet** [reply]
@amyegan So this does not answer the issue raised in the first place where you have a multi-tenants app and you'd like to have `{tenant-name}.pr-number-branch.my-saas.com`, where `tenant-name` can be whatever you want and `pr-number-branch` the unique identifier of a preview app itself.

Correct? Or am I missing something?

**Reply by amyegan** [reply]
Yes and no 😅 Remember that you could only ever use wildcard subdomains on the `*.vercel.app` domain associated with your ***Vercel team slug*** or on your own [custom domain](https://vercel.com/docs/concepts/projects/domains/add-a-domain).

You could assign `*.<user|team slug>.vercel.app` to a preview branch if the goal is to have a wildcard preview environment, but then you would need to add a new custom domain for production deployments. However you cannot add wildcards to [deployment-specific URLs](https://vercel.com/docs/concepts/deployments/generated-urls) or [Git branch URLs](https://vercel.com/docs/concepts/deployments/generated-urls#url-with-git-branch), so you wouldn't be able to have a unique address for each branch or deployment.
